### PR TITLE
Small fixes

### PR DESCRIPTION
**Pull request by kremnev8**
This PR combines a number of small fixes:

1. Fix inherited injected fields not being initialized on derived classes
2. Fix errors injecting a type inheriting a abstract type. I'm not entirely sure why this happens, but the `baseVTablePointer[i].method` in my case was not default, even though all of it's fields were null.
3. Fix delegate type names including assembly names. This might cause very long names and if such name becomes too long (I think 1024 characters or more), the runtime will not like it.
4. Fix NRE setting null to `Value` of `Il2CppReferenceField`.
5. Fix Il2Cpp Detour not being applied if target type is a struct. I have tested this change and it does not cause any issues. Patches get called correctly, and `ref MyStruct __instance` injection works correctly.

I have tested these changes on Core Keeper Unity 2021.3.14.

### CODE REVIEW

File: Il2CppInterop.Runtime/Injection/ClassInjector.cs [hunk 1]
```diff
@@ -128,6 +128,13 @@ public static bool IsTypeRegisteredInIl2Cpp(Type type)
         var currentPointer = Il2CppClassPointerStore.GetNativeClassPointer(type);
         if (currentPointer != IntPtr.Zero)
             return true;
+        if (IsTypeInjectedOnly(type)) return true;
+
+        return false;
+    }
+
+    internal static bool IsTypeInjectedOnly(Type type)
```
Comment (Kasuromi) [inline Il2CppInterop.Runtime/Injection/ClassInjector.cs:136]:
```suggestion
    internal static bool IsManagedTypeInjected(Type type)
```

File: Il2CppInterop.Runtime/Injection/ClassInjector.cs [hunk 2]
```diff
@@ -380,18 +404,18 @@ static void FindAbstractMethods(List<INativeMethodInfoStruct> list, INativeClass
...
             else
             {
                 baseMethod = UnityVersionHandler.Wrap(vTablePointer[i].method);
             }
 
+            if (baseMethod.Name == IntPtr.Zero)
+            {
+                baseMethod = HandleAbstractMethod(i);
+            }
+
```
Comment (Kasuromi) [inline Il2CppInterop.Runtime/Injection/ClassInjector.cs:418]: Could we somehow eliminate this check? Have you tried checking if `methodPtr` is zero?

File: Il2CppInterop.Runtime/DelegateSupport.cs [hunk 3]
```diff
@@ -272,7 +274,7 @@ private static void LogError(string message)
                 throw new ArgumentException(
                     $"Parameter type at {i} has mismatched native type pointers; types: {nativeType.FullName} != {managedType.FullName}");
 
-            if (nativeType.IsByRef || managedType.IsByRef)
+            if (nativeType.IsByRef && managedType.IsByRef)
```
Comment (Kasuromi) [inline Il2CppInterop.Runtime/DelegateSupport.cs:277]: What's the reason for this change?

File: Il2CppInterop.Runtime/Injection/ClassInjector.cs [hunk 4]
```diff
@@ -128,6 +128,13 @@ public static bool IsTypeRegisteredInIl2Cpp(Type type)
         var currentPointer = Il2CppClassPointerStore.GetNativeClassPointer(type);
         if (currentPointer != IntPtr.Zero)
             return true;
+        if (IsTypeInjectedOnly(type)) return true;
```
Comment (Kasuromi) [inline Il2CppInterop.Runtime/Injection/ClassInjector.cs:131]: This check is redundant, injected types always have their native class pointers set in the pointer store.

**Review by Kasuromi — CHANGES_REQUESTED** [review]

File: Il2CppInterop.Runtime/Injection/ClassInjector.cs (hunk 1 above)
Comment (Kasuromi) [reply Il2CppInterop.Runtime/Injection/ClassInjector.cs:136]: We also have `RuntimeSpecificsStore.IsInjected`, you might want to use that instead.

File: Il2CppInterop.Runtime/Injection/ClassInjector.cs (hunk 4 above)
Comment (kremnev8) [reply Il2CppInterop.Runtime/Injection/ClassInjector.cs:131]: I mean I just refactored the method out.

File: Il2CppInterop.Runtime/Injection/ClassInjector.cs (hunk 1 above)
Comment (kremnev8) [reply Il2CppInterop.Runtime/Injection/ClassInjector.cs:136]: Then I would need to get the native class ptr. This feels more convinient.

File: Il2CppInterop.Runtime/Injection/ClassInjector.cs (hunk 4 above)
Comment (Kasuromi) [reply Il2CppInterop.Runtime/Injection/ClassInjector.cs]: Fair enough I'll clean most of the redundant stuff in another commit.

File: Il2CppInterop.Runtime/DelegateSupport.cs (hunk 3 above)
Comment (kremnev8) [reply Il2CppInterop.Runtime/DelegateSupport.cs:277]: I don't actually remember. Probably a stray change. Will revert.

File: Il2CppInterop.Runtime/Injection/ClassInjector.cs (hunk 2 above)
Comment (kremnev8) [reply Il2CppInterop.Runtime/Injection/ClassInjector.cs:418]: Unfortunately, but in my test case it's not zero:
изображение

**Review by Kasuromi — APPROVED** [review]

**Review by ManlyMarco — APPROVED** [review]
